server/test/integration/apiTest/messageClientsTest.js

```diff
@@ -0,0 +1,328 @@
+        heartbeatPhoneNumbers: [],
+        fromPhoneNumber: testFromPhoneNumber,
+        isSendingAlerts: false,
+        isSendingVitals: true,
+      }),
+      await factories.clientDBFactory(db, {
+        displayName: "Test Client Don't Send Vitals",
+        responderPhoneNumbers: [],
+        fallbackPhoneNumbers: [],
+        heartbeatPhoneNumbers: [testNotExpectPhoneNumbers[2]],
+        fromPhoneNumber: testFromPhoneNumber,
+        isSendingAlerts: true,
+        isSendingVitals: false,
+      }),
+      await factories.clientDBFactory(db, {
+        displayName: "Test Client Don't Send",
+        responderPhoneNumbers: [testNotExpectPhoneNumbers[3]],
+        fallbackPhoneNumbers: [testNotExpectPhoneNumbers[3]],
+        heartbeatPhoneNumbers: [testNotExpectPhoneNumbers[3]],
+        fromPhoneNumber: testFromPhoneNumber,
+        isSendingAlerts: false,
+        isSendingVitals: false,
+      }),
+    )
+
+    // create buttons for all non-active clients
+    for (let i = 0; i < 3; i += 1) {
+      testNotExpectResponse.push({
+        to: testNotExpectPhoneNumbers[i + 1],
+        from: testFromPhoneNumber,
+        clientId: nonActiveClients[i].id,
+        clientDisplayName: nonActiveClients[i].displayName,
+      })
+      await buttonDBFactory(db, {
+        clientId: nonActiveClients[i].id,
+        displayName: `Button for ${nonActiveClients[i].displayName}`,
+        buttonSerialNumber: `Button${nonActiveClients[i].id.substr(0, 8)}`,
+        isSendingAlerts: Math.round(Math.random()),
+        isSendingVitals: Math.round(Math.random()),
+      })
+    }
+
+    this.agent = chai.request.agent(server)
+  })
+
+  afterEach(async () => {
+    await db.clearTables()
+    this.agent.close()
+  })
+
+  describe('for a request that uses the correct PA API key, and where Twilio is operating correctly', () => {
+    beforeEach(async () => {
+      // stub twilioHelpers.sendTwilioMessage
+      sandbox.stub(twilioHelpers, 'sendTwilioMessage').returns({ status: 'queued' })
+    })
+
+    afterEach(async () => {
+      sandbox.restore()
+    })
+
+    it('should respond with status 200 (OK)', async () => {
+      const response = await this.agent.post('/api/message-clients').send({ braveKey, message })
+
+      expect(response).to.have.status(200)
+    })
+
+    it('should respond with the same message that was posted', async () => {
+      const response = await this.agent.post('/api/message-clients').send({ braveKey, message })
+
+      expect(response.body.data).to.not.be.undefined
+      expect(response.body.data.message).to.be.a('string').that.equals(message)
+    })
+
+    it('should attempt to message all active clients', async () => {
+      await this.agent.post('/api/message-clients').send({ braveKey, message })
+
+      testExpectPhoneNumbers.forEach(pn => {
+        expect(twilioHelpers.sendTwilioMessage).to.be.calledWith(pn, testFromPhoneNumber, message)
+      })
```

Comment:
Ooooo, interesting one.

Generally speaking, a single test should have a single assertion/expects-statement in it (although there are always exceptions). This is because you don't want to hide the results of later tests if an earlier test fails.

In this case, where the array of `testExpectPhoneNumbers` is `['+11111111111', '+12222222222', '+13333333333', '+14444444444']`, if this test failed on `+11111111111` then we don't know whether it would have failed for all the phone numbers in the array or only some of them. Which could be useful information for fixing the failure.

I think there is a function like `getCalls()` that can be used with stubs like `twilioHelpers.sendTwilioMessage.getCalls()` to get a list or array of all the times it was called (documentation: https://sinonjs.org/releases/v17/spy-call/) and then then you could compare that with the calls that you were expecting and provide a more detailed failure for this kind of test.

Up to you if you want to go back and fiddle with this since this code is already merged. But I'm putting it here for future reference!